cpp/include/raft/sparse/matrix/preprocessing.cuh

```diff
@@ -0,0 +1,446 @@
+                 float b_param = 0.75f);
+  void transform(raft::resources& handle,
+                 raft::device_coo_matrix<ValueType,
+                                         IndexType,
+                                         IndexType,
+                                         IndexType,
+                                         raft::device_uvector_policy,
+                                         raft::PRESERVING> coo_in,
+                 float* results,
+                 bool bm25_on,
+                 float k_param = 1.6f,
+                 float b_param = 0.75f);
+
+ private:
+  void _fit(raft::resources& handle,
+            raft::device_vector_view<IndexType, int64_t> rows,
+            raft::device_vector_view<IndexType, int64_t> columns,
+            raft::device_vector_view<ValueType, int64_t> values,
+            int num_rows);
+  void _fit_feats(IndexType* cols, IndexType* counts, IndexType nnz, IndexType* results);
+  void transform(raft::resources& handle,
+                 raft::device_vector_view<IndexType, int64_t> rows,
+                 raft::device_vector_view<IndexType, int64_t> columns,
+                 raft::device_vector_view<ValueType, int64_t> values,
+                 IndexType nnz,
+                 ValueType* results,
+                 bool bm25_on,
+                 float k_param = 1.6f,
+                 float b_param = 0.75f);
+};
+
+/**
+ * This constructor creates the `SparseEncoder` class with a vocabSize equal to the
+ * int vocab parameter supplied.
+ *
+ * @tparam ValueType
+ *   Type of the values in the sparse matrix.
+ * @tparam IndexType
+ *   Type of the indices associated with the values.
+ *
+ * @param[in] vocab
+ *   Value that represents the number of features that exist for the matrices encoded.
+ */
+template <typename ValueType, typename IndexType>
+SparseEncoder<ValueType, IndexType>::SparseEncoder(int vocab) : vocabSize(vocab)
+{
+  cudaMallocManaged(&featIdCount, vocab * sizeof(int));
+  fullIdLen = 0.0f;
+  numRows   = 0;
+  for (int i = 0; i < vocabSize; i++) {
+    featIdCount[i] = 0;
+  }
+}
+/**
+ * This constructor creates the `SparseEncoder` class with a vocabSize equal to the
+ * int vocab parameter supplied.
+ *
+ * @tparam ValueType
+ *   Type of the values in the sparse matrix.
+ * @tparam IndexType
+ *   Type of the indices associated with the values.
+ *
+ * @param[in] featIdValues
+ *   A map that consists of all the indices and values, to populate the featIdCount array.
+ * @param[in] num_rows
+ *   Value that represents the number of rows observed during fit cycle.
+ * @param[in] full_id_len
+ *   Value that represents the number overall number of features observed during the fit
+ * cycle.
+ * @param[in] vocab_size
+ *   Value that represents the number of features that exist for the matrices encoded.
+ * */
+template <typename ValueType, typename IndexType>
+SparseEncoder<ValueType, IndexType>::SparseEncoder(std::map<int, int> featIdValues,
```

Comment:
Would be good to have a serializer that could export a file that has this information so we can load the fitted class in multiple places.